File: src/main/java/com/bdev/hogwarts_api/service/group/GroupServiceImpl.java
package com.bdev.hogwarts_api.service.group;

import com.bdev.hogwarts_api.dao.GroupDao;
import com.bdev.hogwarts_api.data.converter.group.GroupDtoConverter;
import com.bdev.hogwarts_api.data.converter.group.GroupModelConverter;
import com.bdev.hogwarts_api.data.dto.group.Group;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

import static java.lang.String.format;
import static java.util.stream.Collectors.toList;

@Service
public class GroupServiceImpl implements GroupService {
    @Autowired
    private GroupDao groupDao;

    @Override
    public boolean exists(long id) {
        return groupDao.exists(id);
    }

    @Override
    public List<Group> getAllGroups() {
        return groupDao.findAll().stream().map(GroupModelConverter::convert).collect(toList());
    }

    @Override
    public Optional<Group> getGroup(long id) {
        return Optional.ofNullable(groupDao.getOne(id)).map(GroupModelConverter::convert);
    }

    @Override
    public long createGroup(Group group) {
        if (group.getId() != null) {
            throw new RuntimeException("Group id should be null during creation");
        }

        return groupDao.save(GroupDtoConverter.convert(group)).getId();
    }

    @Override
    public void editGroup(Group group) {
        if (!groupDao.exists(group.getId())) {
            throw new RuntimeException(format("Group with id '%d' does not exist", group.getId()));
        }

        groupDao.save(GroupDtoConverter.convert(group));
    }

    @Override
    public void deleteGroup(long id) {
        if (!groupDao.exists(id)) {
            throw new RuntimeException(format("Group with id '%d' does not exist", id));
        }

        groupDao.delete(id);
    }
}

File: src/main/java/com/bdev/hogwarts_api/service/group/GroupServiceImpl.kt
package com.bdev.hogwarts_api.service.group

import com.bdev.hogwarts_api.dao.GroupDao
import com.bdev.hogwarts_api.data.converter.group.GroupDtoConverter
import com.bdev.hogwarts_api.data.converter.group.GroupModelConverter
import com.bdev.hogwarts_api.data.dto.group.Group
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.lang.String.format

@Service
class GroupServiceImpl : GroupService {
    @Autowired
    private lateinit var groupDao: GroupDao

    override fun getAllGroups(): List<Group> {
        return groupDao
                .findAll()
                .map { GroupModelConverter.convert(it) }
    }

    override fun exists(id: Long): Boolean {
        return groupDao.exists(id)
    }

    override fun getGroup(id: Long): Group? {
        return groupDao.getOne(id)?.let { GroupModelConverter.convert(it) }
    }

    override fun createGroup(group: Group): Long {
        if (group.id != null) {
            throw RuntimeException("Group id should be null during creation")
        }

        return groupDao.save(GroupDtoConverter.convert(group)).id ?: throw RuntimeException()
    }

    override fun editGroup(group: Group) {
        if (!groupDao.exists(group.id)) {
            throw RuntimeException(format("Group with id '%d' does not exist", group.id))
        }

        groupDao.save(GroupDtoConverter.convert(group))
    }

    override fun deleteGroup(id: Long) {
        if (!groupDao.exists(id)) {
            throw RuntimeException(format("Group with id '%d' does not exist", id))
        }

        groupDao.delete(id)
    }
}
